app/src/main/java/com/example/rfid_mobilapp/NfcTagUtil.java

```diff
@@ -31,140 +31,122 @@ public static String getItemId(Intent intent, Activity activity) {
-                for (int i = 0; i < 8; i++) {
-                    cmdRead[10] = (byte) ((offset + i) & 0x0ff);
-                    byte[] response = nfcV.transceive(cmdRead);
-                    Utilities.copyByteArray(response, 1, oldData, i * 4, 4);
-                }
+            byte[] oldData = readBlocks(tagId, nfcV, activity, 0, 8);
 
-                byte[] primeItemId = new byte[16];
-                byte[] primeItemId2 = new byte[16];
-                Utilities.copyByteArray(oldData, 2, primeItemId, 0, 16);
-                if (primeItemId[0] == 1) {
+            byte[] primeItemId = new byte[16];
+            boolean alternativeItemId = false;
 
-                  //  byte[] OptionalBlock = nfcV.transceive(getCommandReadSingleBlock(tagId));
-                //    Utilities.copyByteArray(OptionalBlock, 4, primeItemId2, 0, 16);
+            byte[] primeItemId2 = new byte[16];
+            Utilities.copyByteArray(oldData, 2, primeItemId, 0, 16);
+            if (Utilities.isEmpty(primeItemId)) {
+                return "No Id";
+            }
+            String stringOfPrimaryId = new String(primeItemId, StandardCharsets.UTF_8);
 
-                    noId = false;
-                } else {
-                    noId = Utilities.isEmpty(primeItemId);
-                }
+            if (stringOfPrimaryId.charAt(0) == '1') {
+                alternativeItemId = true;
+                byte[] OptionalBlock = readBlocks(tagId, nfcV, activity, 8, 6);
+                Utilities.copyByteArray(OptionalBlock, 4, primeItemId2, 0, 16);
+            }
+            if (!alternativeItemId) {
+                payloadString = new String(primeItemId, StandardCharsets.UTF_8);
+            } else {
+                payloadString = new String(primeItemId, StandardCharsets.UTF_8) +
+                        new String(primeItemId2, StandardCharsets.UTF_8);
+            }
 
-                if (noId) {
-                    payloadString = "NO ID";
-                } else if (primeItemId[0] != 1) {
-                    payloadString = new String(primeItemId, StandardCharsets.UTF_8);
-                } else {
-                    byte[] newPrimeItemId = new byte[16 * 2];
-
-                    for (int i = 0; i < 16; i++) {
-                        newPrimeItemId[i] = primeItemId[i];
-                    }
-                    for (int i = 0; i < 16; i++) {
-                        newPrimeItemId[i + 16] = primeItemId2[i];
-                    }
-                    payloadString = new String(newPrimeItemId, StandardCharsets.UTF_8);
-                }
-                nfcV.close();
+        }
+        return payloadString;
+    }
 
-            } catch (IOException ioException) {
-                Toast.makeText(activity, "Failed to read tag", Toast.LENGTH_LONG).show();
+    private static byte[] readBlocks(byte[] tagId, NfcV nfcV, Activity activity, int offset, int blocks) {
+        byte[] oldData = new byte[4 * blocks];
```

Comment:
Good work, but I think I would rename `oldData` to `newData`, or `resultData` perhaps.